CODINGSTYLE.md

```diff
@@ -0,0 +1,335 @@
+Braces should be used instead if an argument list has only a lambda, and that lambda does not fit in an inline one-liner.
+
+In general, dot-notation should be used for non-symbolic methods, and infix notation should be used for symbolic methods.
+Infix notation is also used if the only argument is a brace lambda.
+
+Examples:
+
+```scala
+// inline lambda, hence (), hence dot-notation
+list.map(x => x * 2)
+
+// long lambda, hence braces, hence infix notation
+list map { x =>
+  if (x < 5) x
+  else x * 2
+}
+
+// symbolic operator, hence infix notation
+value :: list
+```
+
+When calling a method declared with an empty pair of parentheses, use `()`, except if the method is Java-defined *and* does not have side-effects.
+
+### Method definition
+
+All public and protected methods must have an explicit result type.
+
+Procedure syntax must not be used. `: Unit =` must be used instead.
+
+Side-effect-free methods without formal parameters should be declared without `()`, unless either a) it overrides a methode defined with `()` (such as `toString()`) or b) it implements a Java method in the Java libraries.
+
+### for comprehensions
+
+For-comprehensions must use `()` if they have a single generator without `if`, such as:
+
+```scala
+for (i <- 0 until n)
+  doStuff(i)
+```
+
+Otherwise, it must use `{}`, and there must be one generator per line.
+The body of the `for` is then automatically surrounding by braces too, even if it is a one-liner:
+
+```scala
+for {
+  i <- 0 until n
+  j <- 0 until i
+} {
+  doStuff(i, j)
+}
+```
+
+The `yield` keyword should be placed at the end of the `for` line, followed by braces:
+
+```scala
+for (x <- xs) yield {
+  x * 2
+}
+```
+
+For short `yield` blocks, the following can be used instead:
+
+```scala
+for (x <- xs)
+  yield x * 2
+```
+
+### ScalaDoc
+
+ScalaDoc comments that fit in one line must be written as
+
+```scala
+/** Returns the maximum of a and b. */
+def max(a: Int, b: Int): Int = ???
+```
+
+Multi-line ScalaDoc comments must use the following style:
+
+```scala
+/** Returns the maximum of a and b.
+ *
+ *  If a > b, returns a. Otherwise returns b.
+ */
+def max(a: Int, b: Int): Int = ???
+```
+
+### Non-ScalaDoc comments
+
+Normal comments fitting on one-line should use `//`.
+A comment that does not fit on one line should use the multi-line comment syntax and follow this style:
+
+```scala
+/* This complicated algorithm computes the maximum of two integer values a
+ * and b. If a > b, it computes a, otherwise it computes b.
+ */
+```
+
+
+## Usages of higher-order methods
+
+### Option and js.UndefOr
+
+Use the higher-order methods in the APIs of `Option` and `js.UndefOr` rather than doing pattern matching.
+Note particularly the `fold` method, which should be used instead of the `map`+`getOrElse` combination.
+
+### Collections
+
+Higher-order methods should be favored over loops and tail-recursive methods wherever possible.
```

Comment:
Maybe mention that "possible" includes readability. Tail-recursion should be favored over a ridiculously complex `foldLeft` IMO. (And we do that).